pkg/controllers/metrics/state/node.go

```diff
@@ -0,0 +1,257 @@
+	"github.com/aws/karpenter/pkg/controllers/state"
+	"github.com/aws/karpenter/pkg/utils/resources"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	resourceType    = "resource_type"
+	nodeName        = "node_name"
+	nodeProvisioner = "provisioner"
+	nodePhase       = "phase"
+)
+
+var (
+	allocatableGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "allocatable",
+			Help:      "Node allocatable are the resources allocatable by nodes.",
+		},
+		nodeLabelNames(),
+	)
+
+	podRequestsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_pod_requests",
+			Help:      "Node total pod requests are the resources requested by non-DaemonSet pods bound to nodes.",
+		},
+		nodeLabelNames(),
+	)
+
+	podLimitsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_pod_limits",
+			Help:      "Node total pod limits are the resources specified by non-DaemonSet pod limits.",
+		},
+		nodeLabelNames(),
+	)
+
+	daemonRequestsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_daemon_limits",
+			Help:      "Node total daemon requests are the resource requested by DaemonSet pods bound to nodes.",
+		},
+		nodeLabelNames(),
+	)
+
+	daemonLimitsGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "total_daemon_requests",
+			Help:      "Node total daemon limits are the resources specified by DaemonSet pod limits.",
+		},
+		nodeLabelNames(),
+	)
+
+	overheadGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "nodes",
+			Name:      "system_overhead",
+			Help:      "Node system daemon overhead are the resources reserved for system overhead, the difference between the node's capacity and allocatable values are reported by the status.",
+		},
+		nodeLabelNames(),
+	)
+
+	wellKnownLabels = getWellKnownLabels()
+)
+
+func nodeLabelNames() []string {
+	return append(
+		lo.Values(wellKnownLabels),
+		resourceType,
+		nodeName,
+		nodeProvisioner,
+		nodePhase,
+	)
+}
+
+func forEachGaugeVec(f func(*prometheus.GaugeVec)) {
+	for _, gauge := range []*prometheus.GaugeVec{
+		allocatableGaugeVec,
+		podRequestsGaugeVec,
+		podLimitsGaugeVec,
+		daemonRequestsGaugeVec,
+		daemonLimitsGaugeVec,
+		overheadGaugeVec,
+	} {
+		f(gauge)
+	}
+}
+
+func init() {
+	forEachGaugeVec(func(g *prometheus.GaugeVec) {
+		metrics.Registry.MustRegister(g)
+	})
+}
+
+type NodeScraper struct {
+	cluster       *state.Cluster
+	gaugeLabelMap map[*prometheus.GaugeVec]map[string]prometheus.Labels
```

Comment:
Why the extra layer of indirection. Can't we do 

`gaugeLabelMap map[*prometheus.GaugeVec]prometheus.Labels`